js/src/Bars.ts

```diff
@@ -831,19 +810,23 @@ export class Bars extends Mark {
             return;
         }
 
-        const elements = this.d3el.selectAll(".bargroup").filter((data, index) => {
-            return indices.indexOf(index) !== -1;
-        });
-
         const fill = this.model.get("fill"),
             stroke = this.model.get("stroke"),
             stroke_width = this.model.get("stroke_width");
 
-        elements.selectAll('.bar')
-          .style("fill", fill ? this.get_mark_color.bind(this) : "none")
-          .style("stroke", stroke ? stroke : this.get_mark_color.bind(this))
-          .style("opacity", this.get_mark_opacity.bind(this))
-          .style("stroke-width", stroke_width);
+        this.d3el.selectAll(".bargroup").filter((data, index) => {
+                return indices.indexOf(index) !== -1;
+            })
+            .selectAll('.bar')
+            .style("fill", fill ? this.get_mark_color.bind(this) : "none")
+            .style("stroke", stroke ? stroke : this.get_mark_color.bind(this))
+            .style("opacity", this.get_mark_opacity.bind(this))
+            .style("stroke-width", stroke_width);
+    }
+
+    get_mark_color(data, index) {
+        // Workaround for the bargroup, the color index is not the same as the bar index
+        return super.get_mark_color(data, data.color_index);
```

Comment:
This is the fix. The `index` passed by d3 was wrong in the case of having bars under bargroups, so we need to pass the `color_index` to `get_mark_color` manually.